newsroom/wire/service.py

```diff
@@ -0,0 +1,545 @@
+from typing import Any, cast
+from datetime import datetime, timedelta
+import logging
+
+from bson import ObjectId
+
+from content_api.items.model import PubStatusType
+from superdesk.core.types import Request, Response, SearchRequest, ESQuery, ESBoolQuery
+from superdesk.core import get_app_config
+from superdesk.core.resources import AsyncResourceService
+from superdesk.core.resources.cursor import ElasticsearchResourceCursorAsync
+
+from newsroom.exceptions import AuthorizationError
+from newsroom.types import SectionEnum, Product, TopicResourceModel, UserResourceModel, CompanyResource, WireItem
+from newsroom.auth.utils import get_user_or_none_from_request, get_user_sections
+from newsroom.search.types import NewshubSearchRequest, SearchFilterFunction
+from newsroom.search.base_service import BaseNewshubSearchService
+from newsroom.search.filters import (
+    apply_query_string,
+    apply_date_range,
+    apply_advanced_search,
+    prefill_user,
+    prefill_company,
+    prefill_products,
+    prefill_args_from_topic,
+    apply_section_filter,
+    apply_company_filter,
+    apply_products_filter,
+    validate_request,
+    apply_ids_filter,
+)
+from newsroom.products.products import get_products_by_navigation
+
+from .filters import (
+    WireSearchRequestArgs,
+    default_wire_filters,
+    apply_item_type_filter,
+    apply_filters,
+    apply_embargoed_filters,
+    apply_not_canceled_filter,
+    apply_time_limit_filter,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class WireItemService(AsyncResourceService[WireItem]):
+    async def insert_versioned_document(self, doc_dict: dict[str, Any]):
+        await super().insert_versioned_document(doc_dict)
+
+        if doc_dict.get("pubstatus") == PubStatusType.CANCELLED.value:
+            # If the update is a cancel, we need to cancel all versions
+            await self.mongo_versioned_async.update_many(
+                {
+                    "_id_document": doc_dict["_id_document"],
+                    "pubstatus": {"$ne": PubStatusType.CANCELLED.value},
+                },
+                {"$set": {"pubstatus": PubStatusType.CANCELLED.value}},
+            )
+
+
+class WireSearchServiceAsync(BaseNewshubSearchService[WireSearchRequestArgs, WireItem]):
+    search_args_class = WireSearchRequestArgs
+    filters = default_wire_filters
+    section = SectionEnum.WIRE
+    default_sort = [{"versioncreated", -1}]
+    default_page_size = 25
+
+    def __init__(self):
+        self.service = WireItemService()
+
+    async def get_current_user_bookmarks_count(self, section: SectionEnum | None = None) -> int:
+        user = get_user_or_none_from_request(None)
+        if not user:
+            return 0
+
+        cursor = await self.search(
+            NewshubSearchRequest(
+                section=section or self.section,
+                args=WireSearchRequestArgs(bookmarks=[user.id], page_size=0),
```

Comment:
will this count work as expected if the page_size is equal to 0?

Reply:
Yes, because it uses the hits metadata returned by elastic to [get the count](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/resources/cursor.py#L105).
![image](https://github.com/user-attachments/assets/a64866a1-daad-4610-95a3-394f487fe7b2)